Sprint 5

aldadametov opened the pull request:
1. Логика работы приложения разбита на несколько частей.
2. Работа по генерации вопросов для квиза вынесена в отдельную сущность.
3. Логика показа Алёрта вынесена в отдельную сущность.
4. Реализована функция сохранения лучшего результата store (с проверкой на то, что новый лучший результат сохранён в UserDefaults). К самому алерту в поле message передан сформированный из рекордной игры текст.
5. По окончании раунда вызывается алерт, который соответствует макету из figma

aldadametov committed:
1. Ответственности разбиты по файлам и организованы
2. в MovieQuizViewController.swift были внесены корректировки для его совместной работы с другими файлами проекта.

aldadametov committed:
1. В MovieQuizViewController исправлены ошибки из-за которых билд не проходил корректно
2. QuestionFactory закрыта протоколом
3. MovieQuizViewController стал делегатом для QuestionFactory

aldadametov committed:
Логика показа Алёрта вынесена в отдельную сущность

aldadametov committed:
1. Добавлен файл StatisticService.swift для хранения статистики игр, прописана его логика
2. В message в алёрте теперь выводится более подробная информация со статистикой, которую мы храним c помощью StatisticService.swift

protos17 commented on MovieQuiz/Models/QuizQuestion.swift:
```diff
@@ -0,0 +1,18 @@
+        }
+    
+
+
```
Рекомендую убрать лишние отступы, затем выделить текст и нажать "cmd" + "i" для выравнивания кода

protos17 commented on MovieQuiz/Presentation/AlertPresenter.swift:
```diff
@@ -0,0 +1,24 @@
+//  Created by Алишер Дадаметов on 15.04.2023.
+//
+
+import Foundation
```
При импорте UIKit можно не импортировать Foundation, потому что первый содержит в себе второй

protos17 commented on MovieQuiz/Presentation/AlertPresenter.swift:
```diff
@@ -0,0 +1,24 @@
+import Foundation
+import UIKit
+
+class AlertPresenter {
```
Рекомендую указать класс как final

protos17 commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -130,18 +65,21 @@ final class MovieQuizViewController: UIViewController {
-            let text = "Ваш результат: \(correctAnswers)/10"
+            guard let statisticService = statisticService else { return }
+            statisticService.store(correct: correctAnswers, total: 10)
+            let text = "Ваш результат: \(correctAnswers)/10\n" +
```
Можно использовать многострочный вариант строки

protos17 commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -151,44 +89,50 @@ final class MovieQuizViewController: UIViewController {
+                self.questionFactory?.requestNextQuestion()
+            })
+        
+        AlertPresenter.show(from: self, with: alertModel)
```
Рекомендую презентер вынести в отдельную переменную рядом с фабрикой и обращаться к ней

protos17 commented on MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -0,0 +1,72 @@
+
+import Foundation
+
+class QuestionFactory: QuestionFactoryProtocol {
```
Рекомендую указать как final

protos17 commented on MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -0,0 +1,72 @@
+    private let questions: [QuizQuestion] = [
+            QuizQuestion(
+                image: "The Godfather",
+                text: "Рейтинг этого фильма больше чем 6?",
```
Вопрос везде повторяется, рекомендую избавиться от дублирования кода и сделать вопрос значением по умолчанию для константы text в модели

protos17 commented on MovieQuiz/Services/StatisticService.swift:
```diff
@@ -0,0 +1,93 @@
+    private var _totalAccuracy: Double = 0
+    var totalAccuracy: Double {
+        get {
+            return _totalAccuracy
```
Рекомендую добавить отдельные переменные correct и total, а точность сделать вычисляемым значением из данных переменных и можно не вводить дополнительную - _totalAccuracy

protos17 commented on MovieQuiz/Services/StatisticService.swift:
```diff
@@ -0,0 +1,93 @@
+    private var _gamesCount: Int = 0
+    var gamesCount: Int {
+        get {
+            return _gamesCount
```
Можно не вводить дополнительную переменную, а добавить сохранение в UserDafaults:
userDefaults.integer(forKey: Keys.gamesCount.rawValue)

aldadametov committed:
1.Количество игр и средняя точность теперь хранятся в user defaults
2.Для вычисления средней точности былии введены переменные correct, total
3.Метод store скорректирован соответственно
4.Внесены изменения для ускорения компиляции и чистоты кода

protos17 commented on MovieQuiz/Services/StatisticService.swift:
```diff
@@ -0,0 +1,102 @@
+final class StatisticServiceImplementation: StatisticService {
+    
+    var correct: Int {
+        return userDefaults.integer(forKey: Keys.correct.rawValue)
```
Рекомендую это написать для get

protos17 commented on MovieQuiz/Services/StatisticService.swift:
```diff
@@ -0,0 +1,102 @@
+    
+    var correct: Int {
+        return userDefaults.integer(forKey: Keys.correct.rawValue)
+    }
```
Для set указать:
userDefaults.set(newCorrectCount, forKey: Keys.correct.rawValue)

protos17 commented on MovieQuiz/Services/StatisticService.swift:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    var total: Int {
+        return userDefaults.integer(forKey: Keys.total.rawValue)
```
Здесь сделать как с correct

protos17 commented on MovieQuiz/Services/StatisticService.swift:
```diff
@@ -0,0 +1,102 @@
+    
+    func store(correct count: Int, total amount: Int) {
+        gamesCount += 1
+        let newCorrectCount = userDefaults.integer(forKey: Keys.correct.rawValue) + count
```
Здесь рекомендую обращаться к свойству correct, которое объявили выше на 33 строчке, чтобы не работать напрямую с UserDefaults

protos17 commented on MovieQuiz/Services/StatisticService.swift:
```diff
@@ -0,0 +1,102 @@
+        if GameRecord.newBestRecord(lhs: newGameRecord, rhs: bestGame) {
+            bestGame = newGameRecord
+        }
+        userDefaults.set(newCorrectCount, forKey: Keys.correct.rawValue)
```
Здесь просто указать correct = newCorrectCount

protos17 commented on MovieQuiz/Services/StatisticService.swift:
```diff
@@ -0,0 +1,102 @@
+            bestGame = newGameRecord
+        }
+        userDefaults.set(newCorrectCount, forKey: Keys.correct.rawValue)
+        userDefaults.set(newTotalAmount, forKey: Keys.total.rawValue)
```
Все то же, что и для correct, сделать для total
